### ISSUE TITLE
Use forward calls in Encoder and Decoder derivatives instead of encode() and decode()

### ISSUE THREAD

**Issue by abhishekabhishek** [ezeeEric/DiVAE#15]
According to [nn.Module](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module) documentation, any subclasses of nn.Module should override the `forward()` method however this is not being done in `Network()` and its subclasses https://github.com/ezeeEric/DiVAE/blob/c45666b7d346811be652ca6dff961e6966529e6d/utils/networks.py#L19

![image](https://user-images.githubusercontent.com/28398835/109231644-97276400-778c-11eb-9126-c0482e877b8b.png)

Could this become an issue once we start using GPUs etc. where models are often cloned using the computational graph ?

**Comment by ezeeEric** [comment]
The `network`, `encoder` and `decoder` derivatives we create from `nn.Module` are never used directly, no forward call is performed.

The place to look at is in the models files, specifically the [autoencoderbase](https://github.com/ezeeEric/DiVAE/blob/c45666b7d346811be652ca6dff961e6966529e6d/models/autoencoderbase.py). Here we inherit from `nn.Module` and overwrite the `forward` method. The method is overwritten for each new model.

**Comment by abhishekabhishek** [comment]
Hmm. The `encode()`, `decode()` methods from the subclasses `network`, `encoder` and `decoder` are still being called in the `forward()` method for the models e.g. autoencoder. These should technically should be `forward()` methods and not object methods since data is being passed though the layers.

**Comment by ezeeEric** [comment]
Currently, the forward method in Network, Encoder, Decoder etc. are all pointing to pytorch's [default implementation ](https://pytorch.org/docs/stable/_modules/torch/nn/modules/module.html#Module) in `nn.Module`, which is:
```
forward: Callable[..., Any] = _forward_unimplemented
```
which in turn is defined as:
```
def _forward_unimplemented(self, *input: Any) -> None:
[...]
    raise NotImplementedError
```
If that function ever were invoked in what we are doing, we would get an error.

What is the difference between "forward" and "object" methods? The point of subclassing nn.Module in Network() is to broadcast the parameters of that class to pytorch.=

**Comment by ezeeEric** [comment]
I changed the header and reopened this issue, because there is one point to be made: for readabilty reasons we could replace the encode() and decode() definition for Encoder and Decoder with forward() calls. I do not think this has any impact on performance, hence I label it with `prettify` ;)
